Add custom response pages, redirect back option.

TomAbel opened the pull request:
On install create new content type.
Makes one success and failure page of that content type.
Changes success/failure URL from textfield to drop down select.
Add option of setting success/failure URL to redirect back.

Commit:
Add custom response pages, redirect back option.

On install create new content type.
Makes one success and failure page of that content type.
Changes success/failure URL textfield to drop down select.
Add option of setting success/failure URL to redirect back.

kaittodesk commented on smaily_for_drupal/config/install/core.entity_form_display.node.smly_page.default.yml:
```diff
@@ -0,0 +1,29 @@
+      settings:
+        display_label: true
+      third_party_settings: {  }
+      type: boolean_checkbox
```
I recommend making this field into a select box (let's say `response_page_type`) where you can choose between "Success page" or "Failure page" options. This keeps the door open for us to add new page types more easily if necessary.

TomAbel replied:
Drupal doesn't have select in field type when creating it manually

kaittodesk commented on smaily_for_drupal/config/install/core.entity_form_display.node.smly_page.default.yml:
```diff
@@ -0,0 +1,29 @@
+    body:
+      label: hidden
+      type: text_textarea_with_summary
+      weight: 101
```
I recommend having `weight` values in the step of (at least) 10, e.g. 110, 120, etc. That will make adding new fields between existing ones easier.

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -28,3 +29,57 @@ function smaily_for_drupal_help($route_name, RouteMatchInterface $route_match) {
       return $output;
   }
 }
+function smaily_for_drupal_install() {
```
Formatting: missing spacing between functions.

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -28,3 +29,57 @@ function smaily_for_drupal_help($route_name, RouteMatchInterface $route_match) {
+    'format' => 'basic_html',
+  ];
+  $node = Node::create([
+    'type' => 'smly_page',
```
I feel the type value is not descriptive enough. How about `smaily_response_page`? Or is there a length limit for the value?

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -28,3 +29,57 @@ function smaily_for_drupal_help($route_name, RouteMatchInterface $route_match) {
+  // Delete entities.
+  if (!empty($nids)) {
+    $entities = $nodeStorage->loadMultiple($nids);
+    $nodeStorage->delete($entities);
```
Will this remove `smly_page` content type also or does Drupal 8 handles that out of the box through installation configuration files?

TomAbel replied:
Since I specified module: required: Drupal can handle that itself

kaittodesk replied:
Cool.

kaittodesk commented on smaily_for_drupal/src/Form/SubscribeForm.php:
```diff
@@ -117,4 +120,38 @@ public function removeHiddenDrupalInputs(array $form) {
+   * Handle and display response messages from Smaily.
+   */
+  public function handleResponseMessage() {
+    $request = \Drupal::request();
```
There isn't anything wrong with this code technically, but it can be simplified so you wouldn't need to access `query->all()` all the time.

```php
$params = \Drupal::request()->query->all();
$message = isset($params['message']) ? trim($params['message']) : NULL;
$code = isset($params['code']) ? (int) trim($params['code']) : NULL;

if ($message && $code) {
    // ...
}
```

And to reduce the **curly-bracket-pyramid-of-doom** you can always return early in a method:
```php
if (!$message && !$code) {
    return;
}

switch($code) {
    // ...
}
```

kaittodesk commented on smaily_for_drupal/src/Form/SubscribeForm.php:
```diff
@@ -76,6 +78,7 @@ public function buildForm(array $form, FormStateInterface $form_state) {
     // Remove token so as to not send it to Smaily as a field.
     $form['#token'] = FALSE;
     $form['#after_build'][] = [$this, 'removeHiddenDrupalInputs'];
+    $this->handleResponseMessage();
```
This will be executed only when you have a form on your page.

TomAbel replied:
This is correct. Messages are shown when you are redirected back, otherwise the success/failure page will give the user feedback. I can't imagine how to create a site-wide listener and wouldn't it be overkill for this module?

kaittodesk replied:
Maybe you can use [hook_preprocess](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Render%21theme.api.php/function/hook_preprocess/8.2.x) to show success/failure page response message?

TomAbel replied:
I think hook_preprocess is used for themes only. I would need a .theme file for that.

kaittodesk replied:
`hook_preprocess` works with modules too. `smaily_for_drupal_preprocess` would be executed for each visible block. But I guess `hook_page_attachments` works too.

Commit:
rename smly_page, refactor handlemessages

Commit:
replaced success boolean with radio

Commit:
renamed field_success_page to response_page_type

kaittodesk commented on smaily_for_drupal/config/install/field.field.node.smaily_response_page.body.yml:
```diff
@@ -0,0 +1,25 @@
+entity_type: node
+bundle: smaily_response_page
+label: Body
+description: 'Uncheck <em>success page</em> to set as failure page.'
```
This description is no longer valid and can be removed.

kaittodesk commented on smaily_for_drupal/config/install/field.field.node.smaily_response_page.field_response_page_type.yml:
```diff
@@ -0,0 +1,21 @@
+field_name: field_response_page_type
+entity_type: node
+bundle: smaily_response_page
+label: 'Success page'
```
This label is not descriptive enough. Use something like "Choose response page type".

Commit:
Added PHP cURL troubleshooting help

Commit:
Add hook_page_attachments for error handling

Moved handleResponseMessages() to hook_page_attachments
Hook will be run on loading every page.
Previously it would run only when the form was also loaded.

Commit:
remove invalid description, more descriptive label
